### PR TITLE
make logger configurable

### LINKED ISSUE #81: Make lnmux logging suitable for prod environment
Currently in lnmux we use the default zap development configuration:

`config := zap.NewDevelopmentConfig()`

This is not suitable for a prod environment and we need to make it configurable (for example, log encoding must be in `json` for a prod environment otherwise severity will not be displayed correctly)

We could for example add in [this](https://github.com/bottlepay/lnmux/pull/80)  `LoggingConfig`  some fields like `Level`, `Format` (console or json), `withEncodeCaller`,...

### PR DESCRIPTION
**Pull request by Torakushi**
close #81

### CODE REVIEW

File: cmd/lnmuxd/config.go [hunk 1]
```diff
@@ -39,11 +39,22 @@ type Config struct {
...
+	// Level defined the minimum log level to be output
+	Level string `yaml:"level"`
+
+	// Format defined the output format:
+	//  'console' will be human-readable
+	//  'json' will be machine-readable.
+	Format string `yaml:"format"`
+
+	// WithCaller indicates if the caller is displayed in the log
+	WithCaller bool `yaml:"withCaller"`
```
Comment (Torakushi) [inline cmd/lnmuxd/config.go:58]: I used this field to keep the actual behaviour as it is.

But if it is not a problem to display the caller in the console, then I will remove it

File: lnmux.yml.example [hunk 2]
```diff
@@ -18,3 +18,9 @@ db:
   # maxConnAge: "15s"
   # poolTimeout: "15s"
   # idleTimeout: "15s"
+logging:
+  # level: "debug"
+  # format: "console"
+  # withCaller: true
+  # grpcLogging: true
+  # grpcLogging: true
```
Comment (aakselrod-nydig) [inline lnmux.yml.example]: This is duplicated

File: cmd/lnmuxd/config.go (hunk 1 above)
Comment (aakselrod-nydig) [reply cmd/lnmuxd/config.go:58]: I like having the option.

**Review by aakselrod-nydig — COMMENTED** [review]
Looks good, just a duplicated line in the example config.

File: lnmux.yml.example [hunk 3]
```diff
@@ -18,3 +18,9 @@ db:
   # maxConnAge: "15s"
   # poolTimeout: "15s"
   # idleTimeout: "15s"
+logging:
+  # level: "debug"
+  # format: "console"
+  # withCaller: true
+  # grpcLogging: true
+  # grpcPayloadLogging: true
```
Comment (aakselrod-nydig) [inline lnmux.yml.example:26]: ⚡

File: cmd/lnmuxd/log.go [hunk 4]
```diff
@@ -1,16 +1,74 @@
...
+	encoderConfig.TimeKey = "time"
+
+	if !withCaller {
+		encoderConfig.EncodeCaller = nil
+	}
+
+	// format is 'console' by default
+	if format == "" {
+		// Add color to the level
+		encoderConfig.EncodeLevel = zapcore.CapitalColorLevelEncoder
```
Comment (aakselrod-nydig) [inline cmd/lnmuxd/log.go:47]: ⚡

**Review by aakselrod-nydig — APPROVED** [review]
LGTM

File: README.md [hunk 5]
```diff
@@ -21,6 +21,7 @@ For the setup, it is assumed that there are multiple LND nodes running with conn
   * LND nodes configuration: TLS certificate, macaroon, address and pubkey. Pubkey is configured as a protection against unintentionally connecting to the wrong node.
   * Postgres connection string for the database created above (can be overridden by the environment variable `LNMUX_PERSISTENCE_POSTGRES_DSN` to avoid exposing credentials).
   * 32-byte Identity PRIVATE key. This key is used to generate invoices and decode incoming htlcs. In a production environment, this key must be protected carefully (can be overridden by the environment variable `LNMUX_IDENTITY_KEY` to avoid exposing the private key).
+  * Your logging configuration. You can choose the level, the encoding format, to display grpc logs or not,...
```
Comment (joostjager) [inline README.md]: "or not,..." ?

File: cmd/lnmuxd/log.go [hunk 6]
```diff
@@ -1,16 +1,74 @@
...
+		OutputPaths:      []string{"stderr"},
+		ErrorOutputPaths: []string{"stderr"},
+	}
+
+	lo, err := config.Build()
+	if err != nil {
+		return err
+	}
+
+	lo = lo.With(zap.String("svc", "lnmux"))
```
Comment (joostjager) [inline cmd/lnmuxd/log.go]: Isn't this unnecessary as there are no other services in this process?

**Review by joostjager — COMMENTED** [review]
Nice addition

File: README.md (hunk 5 above)
Comment (Torakushi) [reply README.md]: Maybe it is unclear, I will remove the 'or not'

File: cmd/lnmuxd/log.go (hunk 6 above)
Comment (Torakushi) [reply cmd/lnmuxd/log.go]: It is interesting to have this, so we can filter logs given the service name (in an environment where we have various services).

Comment (joostjager) [reply cmd/lnmuxd/log.go]: Isn't a field for the service name automatically added in such environments by the environment itself? I mean, in k8s for example, not all logs line of all processes are piled on a single heap with any additional identification of where it originates from?

File: README.md (hunk 5 above)
Comment (joostjager) [reply README.md]: Ah, you mean "etcetera"? Maybe more informative to write it as something like "the most important options are x and y".

Comment (Torakushi) [reply README.md]: Yeah, I mean etcetera, but you are right, your formulation looks better!

File: cmd/lnmuxd/log.go (hunk 6 above)
Comment (Torakushi) [reply cmd/lnmuxd/log.go]: Yeah, I guess we can filter per container name in k8s. But maybe we will have some other environment where the service name will be missing? 
I don't really know about other env.
 K8s remains the norm so I can remove it I guess!

**Review by joostjager — APPROVED** [review]

**Pull request merged**